spec/Section 6 -- Execution.md

```diff
@@ -552,9 +552,30 @@ set or coercing a scalar value.
 ExecuteField(objectType, objectValue, fieldType, fields, variableValues):
   * Let {field} be the first entry in {fields}.
   * Let {fieldName} be the field name of {field}.
+  * Let {requiredStatus} be the required status of {field}.
   * Let {argumentValues} be the result of {CoerceArgumentValues(objectType, field, variableValues)}
   * Let {resolvedValue} be {ResolveFieldValue(objectType, objectValue, fieldName, argumentValues)}.
-  * Return the result of {CompleteValue(fieldType, fields, resolvedValue, variableValues)}.
+  * Let {modifiedFieldType} be {ModifiedOutputType(fieldType, requiredStatus)}.
+  * Return the result of {CompleteValue(modifiedFieldType, fields, resolvedValue, variableValues)}.
+
+## Accounting For Client Controlled Nullability Designators
+
+A field can have its nullability status set either in its service's schema, or
+it can be overriden by a designator (! or ?) for the duration of an execution.
+In order to determine a field's true nullability, both are taken into account
+and a final type is produced.
+
+ModifiedOutputType(outputType, requiredStatus):
+  * If {requiredStatus} is 'required' and {outputType} is not a Non-Nullable type:
+    * Return `Non-Null` with an inner type of {outputType}.
```

Comment:
This does not consider list designators like [!]!

The algorithm is recursive and must pass on the inner type to `ModifiedOutputType`.

I have implemented now a version that covers all cases in HotChocolate. Do you want me to rewrite the `ModifiedOutputType`  spec?

Reply:
It's not fully polished yet bit here is the C# algorithm for this:
https://github.com/ChilliCream/hotchocolate/blob/d053aedd28aaf535732a650c25bd82cce77dc47b/src/HotChocolate/Core/src/Types/Types/Extensions/TypeExtensions.cs#L524

Reply:
Its not fully polished yet, but here is our C# implementation of the algorithm:
https://github.com/ChilliCream/hotchocolate/blob/d053aedd28aaf535732a650c25bd82cce77dc47b/src/HotChocolate/Core/src/Types/Types/Extensions/TypeExtensions.cs#L524